File: implementations/python/src/mzprov/paths.py
"""Path conventions shared between the signer and the verifier.

The signer copies the user's config bytes to a conventional location;
the verifier reads from the same location to recompute the signed
``config_hash``. Both sides MUST derive the path from the on-disk
artifact (or sidecar) location alone — never from a payload field
like ``experiment_name`` — because payload fields are
attacker-controllable and a tampered field could otherwise redirect
the verifier to a phantom file. Centralizing the conventions here
keeps that property in one place.

Three conventions:

- **JSON sidecar transport.** The config copy sits at
  ``{sidecar-stem}.config.toml`` next to the sidecar, where the
  sidecar stem is the sidecar's basename minus ``.provenance.json``.
- **Embedded .d transport.** The config copy sits at
  ``{d-stem}.config.toml`` next to the .d, where the .d stem is the
  directory name minus the ``.d`` suffix.
- **Embedded mzML transport.** The config copy sits at
  ``{mzml-stem}.config.toml`` next to the mzml file.
"""

from __future__ import annotations

from pathlib import Path
from typing import Union

PathLike = Union[str, Path]


def sidecar_config_path(sidecar_path: PathLike) -> Path:
    """Conventional config-copy path for the JSON sidecar transport.

    For ``foo.provenance.json`` returns ``foo.config.toml`` in the
    same directory. The sidecar suffix ``.provenance.json`` is
    stripped explicitly so a sidecar like ``a.b.provenance.json``
    yields ``a.b.config.toml``, not ``a.config.toml``.
    """
    sidecar_path = Path(sidecar_path)
    name = sidecar_path.name
    if name.endswith(".provenance.json"):
        stem = name[: -len(".provenance.json")]
    else:
        stem = sidecar_path.stem
    return sidecar_path.parent / f"{stem}.config.toml"


def embedded_d_config_path(d_path: PathLike) -> Path:
    """Conventional config-copy path for the embedded-d transport.

    For ``…/sample.d`` returns ``…/sample.config.toml``. The ``.d``
    suffix is stripped explicitly; if the directory does not end in
    ``.d`` the full directory name is used as the stem.
    """
    d_path = Path(d_path)
    name = d_path.name
    if name.endswith(".d"):
        stem = name[: -len(".d")]
    else:
        stem = name
    return d_path.parent / f"{stem}.config.toml"


def embedded_mzml_config_path(mzml_path: PathLike) -> Path:
    """Conventional config-copy path for the embedded-mzml transport.

    For ``…/sample.mzML`` returns ``…/sample.config.toml``.
    """
    mzml_path = Path(mzml_path)
    return mzml_path.with_name(mzml_path.stem + ".config.toml")

File: implementations/csharp/Mzprov/Paths.cs
using System;
using System.IO;

namespace Mzprov;

// Config-copy path conventions shared between signing and verification.
// Both sides MUST derive the config path from the on-disk artifact/sidecar
// location alone (never from a payload field). Mirrors mzprov.paths.
internal static class Paths
{
    public static string SidecarConfigPath(string sidecarPath)
    {
        string name = BaseName(sidecarPath);
        string stem = name.EndsWith(".provenance.json", StringComparison.Ordinal)
            ? name.Substring(0, name.Length - ".provenance.json".Length)
            : Path.GetFileNameWithoutExtension(name);
        return Path.Combine(DirName(sidecarPath), stem + ".config.toml");
    }

    public static string EmbeddedDConfigPath(string dPath)
    {
        string name = BaseName(dPath);
        string stem = name.EndsWith(".d", StringComparison.Ordinal) ? name.Substring(0, name.Length - 2) : name;
        return Path.Combine(DirName(dPath), stem + ".config.toml");
    }

    public static string EmbeddedMzmlConfigPath(string mzmlPath) =>
        Path.Combine(DirName(mzmlPath), Path.GetFileNameWithoutExtension(BaseName(mzmlPath)) + ".config.toml");

    public static string BaseName(string path) =>
        Path.GetFileName(path.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar));

    public static string DirName(string path) =>
        Path.GetDirectoryName(path.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)) ?? ".";
}
